### DISCUSSION TITLE
General discussion about the library

### DISCUSSION THREAD

**Discussion by Lugaresi** [Lugaresi/codesys2-mqtt-library#5] [General]
Bello everyone,
 Since both issues are fixed now, and thank you very much for that, I guess we can use discussions to keep an open channel of communication between users to talk about generic implementation, feature requests, feedback and whatnot.

I'm tagging the people who were in the previous issues so you know, feel free to not RSVP.
Happy coding and merry Christmas!
   Flavio

@huwylphi @cholokla @ddokupil @chillymattster

**Comment by ddokupil** [comment]
Excellent idea! I was to ask you for your implementations, especially regarding HASS. For now I have a very simple implementation in a separate program. It just sends all my MX registers to HASS once a second and in HASS I have automations that parse them into binary sensors. The other way is HASS setting the same MX registers via http WRITEPI requests. It works but is far from what i'd like to have.
I'd like to be able to send MQTT messages once light state changes (and send update on all lights when HASS sends "online" on homeassistant/status topic), and make it work the other way too with HASS sending messages to MQTT and Wago reacting to them accordingly. Wold be easier to me if it was python but in codesys I have to gain a lot more experience to be able to implement such things on my own. I hope we can figure out something sophisticated together.

**Reply by huwylphi** [reply]
Hi @ddokupil,
I think what you're wishing to achieve between you PLC and HASS is achievable with that library (I'm doing more or less the same, just not (yet) with HASS).
In your CodeSys program you can:
1. connect the MQTT client to your broker and
2. then subscribe to some MQTT topic
3. then catch new messages published by the broker on that topic by checking the output variable _q_xReceivedMessageNotification_ of the MQTT client function block instance.  
_q_xReceivedMessageNotification_ will be _TRUE_ just one PLC cycle. When that output is _TRUE_ you can parse the message's content and consume its data as your PLC would need.
4. For publishing data from your PLC to the MQTT broker, you might not need to do that periodically (every seconds). If that is the case, you could monitor changes of your PLC data you want to publish and only publish when they change.  
For example for a _boolean_ value, you can use an _R_TRIG_ function block. Or you can also use some nice function blocks from some well known [OSCAT library](http://www.oscat.de/en/component/jdownloads/download/2-oscat-basic/3-oscat-basic-333.html) like [B_TRIG and A_TRIG](http://www.oscat.de/en/component/jdownloads/send/2-oscat-basic/9-oscat-basic333-en.html?Itemid=0) that will set an output _Q_ to _TRUE_ just one cycle if the input value changed.  
You can use these function blocks to monitor change of value and then trigger publishing of your PLC data to the MQTT broker.
5. And regarding the HASS status, your PLC could simply subscribe to the _homeassistant/status_ topic
6. then like for point 3 above monitor for "online" message
7. that would trigger the publishing of all PLC data that HASS would need to get update from when getting online.

I hope my ideas will help you.
Kind regards
Philippe

**Reply by cholokla** [reply]
> Hi @ddokupil, I think what you're wishing to achieve between you PLC and HASS is achievable with that library (I'm doing more or less the same, just not (yet) with HASS). In your CodeSys program you can:
> 
> 1. connect the MQTT client to your broker and
> 2. then subscribe to some MQTT topic
> 3. then catch new messages published by the broker on that topic by checking the output variable _q_xReceivedMessageNotification_ of the MQTT client function block instance.
>    _q_xReceivedMessageNotification_ will be _TRUE_ just one PLC cycle. When that output is _TRUE_ you can parse the message's content and consume its data as your PLC would need.
> 4. For publishing data from your PLC to the MQTT broker, you might not need to do that periodically (every seconds). If that is the case, you could monitor changes of your PLC data you want to publish and only publish when they change.
>    For example for a _boolean_ value, you can use an _R_TRIG_ function block. Or you can also use some nice function blocks from some well known [OSCAT library](http://www.oscat.de/en/component/jdownloads/download/2-oscat-basic/3-oscat-basic-333.html) like [B_TRIG and A_TRIG](http://www.oscat.de/en/component/jdownloads/send/2-oscat-basic/9-oscat-basic333-en.html?Itemid=0) that will set an output _Q_ to _TRUE_ just one cycle if the input value changed.
>    You can use these function blocks to monitor change of value and then trigger publishing of your PLC data to the MQTT broker.
> 5. And regarding the HASS status, your PLC could simply subscribe to the _homeassistant/status_ topic
> 6. then like for point 3 above monitor for "online" message
> 7. that would trigger the publishing of all PLC data that HASS would need to get update from when getting online.
> 
> I hope my ideas will help you. Kind regards Philippe

Hi,
that's exactly what I've done in my tests.
I'm just a beginner in the PLC area but it's working properly at least on the bench.
For now I'm publishing all the changes of the outputs and subscribing for the messages from external sources to change output states.
To hold output values I use intermediary variable _aOutputs_ (which can be set through the modbus protocol either) and in "TASK_OUTPUTS" I'm writing it's values straight to the DO variables.

aOutputs AT %MB144 : ARRAY[ARRAY_MIN..NUM_OUTPUTS] OF BOOL;

![wago_mqtt_outputs](https://user-images.githubusercontent.com/15654069/209582659-b98c742e-190d-4502-8444-d82f07951804.png)

Greetings,
robert

**Comment by ddokupil** [comment]
Well, your ideas are obvious. I will put publishing data from HASS to PLC aside. I started from improving sending status from Wago to HASS.
I created my custom LatchingRelay FB with following implementation:
```
FUNCTION_BLOCK HASS_LIGHT
VAR_INPUT
	xPhysSwitch: BOOL := FALSE;
	xVirtSwitch: BOOL := FALSE;
	xCentOFF: BOOL := FALSE;
	xCentON: BOOL := FALSE;
	bModeOnReset: BYTE := 3;
END_VAR
VAR_OUTPUT
	xActuator: BOOL;
END_VAR
VAR
	LATCH: Fb_LatchingRelay;
	RISE: R_TRIG;
	FALL: F_TRIG;
END_VAR
```
```
LATCH(xSwitch:=xPhysSwitch);
RISE(CLK:=xVirtSwitch);
FALL(CLK:=xVirtSwitch);
LATCH(xCentOFF:=FALL.Q OR xCentOff);
LATCH(xCentON:=RISE.Q OR xCentOn);
LATCH(bModeOnReset:=bModeOnReset);
xActuator:=LATCH.xActuator;
IF (xPhysSwitch OR xCentOff OR xCentOn)
THEN
	PLC_PRG.HASS_UPDATE_STATUS;
END_IF
```
it calls PLC_PRG.HASS_UPDATE_STATUS whis is 
```
(...)
HASS_PRYSZ_LAZ_M:=PRYSZ_LAZ_M;
HASS_PRYSZ_LAZ_D:=PRYSZ_LAZ_D;
IF HASS_SW_PRALN<>SW_PRALN
THEN
	sMQTTTopic:=WAGO_LIGHT_TOPIC;
	sMQTTMessage:='{"HASS_SW_PRALN": ';
	IF SW_PRALN=TRUE
	THEN
		sMQTTMessage:=CONCAT(sMQTTMessage, '"on"}');
	ELSE
		sMQTTMessage:=CONCAT(sMQTTMessage, '"off"}');
	END_IF;
	xSendPayload:=TRUE;
	HASS_SW_PRALN:=SW_PRALN;
END_IF;
HASS_WC_LAZ_M_H:=WC_LAZ_M_H;
(...)
```
So first two and last line is enough if I read and write status using modbus or http requests. The IF statement is an attempt to update with MQTT.
My MQTT implementation is actually pretty simple:
```
PROGRAM MQTT
VAR
	Client:FB_MQTTClient;
END_VAR
```
```
Client(i_sUsername:='user', i_sPassword:='passwd', i_xEnable:=TRUE, i_sBrokerAddress:='192.168.1.4', i_tCommTimeout:=t#10s, i_tCommTimeout:=t#1m, i_sPublishTopic:=sMQTTTopic, i_sPublishMessage:=sMQTTMessage, i_xPublish:=xSendPayload);
xSendPayload:=FALSE;
```
...and it actually works... as long as I don't change the state too frequently. If I hit the switch more than once a second it tends to ge lost with the updates. I assume MQTT needs up to a second to push the message and be ready to receive another one... Too slow. I can include many states at once within one topic and it would work if I for example change whole scene but If I hit the same switch twice too quick I will most likely end up with wrong state in HASS. I had the same problem with http implementation instead of mqtt.
So that kind of problems to solve...
I will need to somehow check the last sent state and retry until they match, or come up with something else...

**Comment by huwylphi** [comment]
Hello @ddokupil, @cholokla 

just wanted to add one remark regarding updating HA MQTT entities when it comes online again (e.g. after a restart).
In a previous post I mentioned that the PLC could subscribe to the _homeassistant/status_ topic for monitoring the _online_ status and then publish (refresh) all entities states. But this might actually not be necessary since MQTT has a feature for that. When the PLC is publishing data on a topic, it can set the RETAIN flag (input var _i_xRetain_). That way when a new subscription is established by HA (this will happen when it comes online again), the last retained message, if any, on each topic will be sent by the broker to HA. This is one advantage of MQTT since the publisher (here the PLC) doesn't need to care about subscribers (here HA) and delegates the role of keeping last values / messages to the MQTT broker.

Then regarding how fast MQTT messages can be exchanged between publisher and subscribers, this could depend on several things like the infrastructure, network, resources of the PLC and HA hosts, or PLC runtime cycles. MQTT is actually very lightweight, so I guess it should be fast; probably within the hundred milliseconds range.
